Add Order model

dmarmul opened the pull request:
Added Order and OrderItem to project. Implement all needed classes for it.

dmarmul committed:
Added Order and OrderItem to project. Implement all needed classes for it.

liliia-ponomarenko commented on src/main/java/org/example/bookshop/controller/CartController.java:
```diff
@@ -41,7 +41,7 @@ public ShoppingCartDto get(@AuthenticationPrincipal User user) {
     @Operation(summary = "Add books to cart",
             description = "Add books to the shopping cart."
-                    + " Fields bookId and quantity can't must be not less than 1, can't be null.")
+                    + " Fields bookId and quantity must be positive, can't be null.")
```
in programming, 0 is considered a positive number)

liliia-ponomarenko commented on src/main/java/org/example/bookshop/controller/OrderController.java:
```diff
@@ -0,0 +1,80 @@
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Get orders history",
+            description = "Retrieve all user's orders")
+    public Set<OrderDto> getAll(@AuthenticationPrincipal User user,
```
better return Page

liliia-ponomarenko commented on src/main/java/org/example/bookshop/service/OrderService.java:
```diff
@@ -0,0 +1,23 @@
+import org.springframework.data.domain.Sort;
+
+public interface OrderService {
+    OrderDto add(@Valid OrderPurchaseRequestDto requestDto, User user);
+
+    Set<OrderDto> getAll(User user, Sort sort, Pageable pageable);
+
+    void update(@Valid OrderStatusRequestDto requestDto, Long orderId, User user);
```
validation on the controller layer is enough

liliia-ponomarenko commented on src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,134 @@
+        ShoppingCart shoppingCart = cartRepository.findByUserId(
+                user.getId()).orElseThrow(() -> new EntityNotFoundException(
+                "Can't find order for user: " + user.getEmail()));
+        Order order = orderMapper.toEntity(shoppingCart);
+        order.setShippingAddress(requestDto.getShippingAddress());
+        order.setOrderDate(LocalDateTime.now());
+        order.setStatus(Order.Status.PENDING);
+        Set<OrderItem> orderItems = findOrderItems(user);
+        orderItems.forEach(item ->
+                item.setPrice(item.getPrice().multiply(new BigDecimal(item.getQuantity()))));
+        order.setTotal(getTotalPrice(orderItems));
```
better move this logic to the private method

liliia-ponomarenko commented on src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,134 @@
+    private final OrderMapper orderMapper;
+
+    @Override
+    public OrderDto add(OrderPurchaseRequestDto requestDto, User user) {
```
Let’s  use Transactional here ;)

liliia-ponomarenko commented on src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,134 @@
+    public void update(OrderStatusRequestDto requestDto, Long orderId, User user) {
+        Order order = findOrderByIdAndUserId(orderId, user.getId());
+        Order.Status status = Order.Status.valueOf(requestDto.getStatus());
+        if (order.getStatus().equals(status)) {
+            throw new DataProcessingException("Order already has status: " + status);
+        }
```
I don’t see your global exception handler but I guess it provide internal server error. is it ok?

dmarmul replied:
There is provides 401 http response

liliia-ponomarenko replied:
Okay, your filter chain does it. Anyway, you see an exception in the console. Is it ok? 
401 Unauthorized  - is it the appropriate status for this exception? 
You should handle this exception and return CONFLICT.

liliia-ponomarenko commented on src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,134 @@
+        return findOrderItems(user).stream()
+                .map(orderItemMapper::toDto)
+                .collect(Collectors.toSet());
+        // Have alternative solve, it should be faster,
+        // but need 1 more method in orderItemRepository
+        // return orderItemRepository.findAllByOrderId(orderId)
+//                .stream()
+//                .filter(Optional::isPresent)
+//                .map(Optional::get)
+//                .map(orderItemMapper::toDto)
+//                .collect(Collectors.toSet());
```
Let’s  remove the commented code

liliia-ponomarenko commented on src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,134 @@
+        return orderItemMapper.toDto(itemOptional.get());
+    }
+
+    private Order findOrderByIdAndUserId(Long orderId, Long userId) {
+        Optional<Order> optionalOrder = orderRepository.findByIdAndUserId(orderId, userId);
+        if (optionalOrder.isEmpty()) {
+            throw new EntityNotFoundException("Can't find order with id: " + orderId);
+        }
+        return optionalOrder.get();
+    }
```
```suggestion
    private Order findOrderByIdAndUserId(Long orderId, Long userId) {
        return orderRepository.findByIdAndUserId(orderId, userId).orElseThrow(() -> 
                 new EntityNotFoundException("Can't find order with id: " + orderId)
        );
    }
```

liliia-ponomarenko commented on src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,134 @@
+    @Override
+    public OrderItemDto getOrderItem(Long orderId, Long itemId, User user) {
+        findOrderByIdAndUserId(orderId, user.getId());
+        Optional<OrderItem> itemOptional = orderItemRepository.findByIdAndOrderId(itemId, orderId);
+        if (itemOptional.isEmpty()) {
+            throw new EntityNotFoundException("Can't find item with id: " + itemId);
+        }
```
Let’s  use `orElseThrow`

liliia-ponomarenko commented on src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,134 @@
+                .stream()
+                .map(orderItemMapper::toEntity)
+                .collect(Collectors.toSet());
+        if (orderItem.isEmpty()) {
+            throw new EntityNotFoundException(
+                    "Can't find order items for user: " + user.getEmail());
+        }
```
why should it occur Exception?

dmarmul replied:
This check doesn't allow to place an empty order. Added more description into exception

liliia-ponomarenko reviewed (changes requested):
Good job! Let’s improve your solution ;)

dmarmul committed:
Return Page from getAll method.
Used Transactional in add method, replace logic of creating order in private method.
Didn't replace DataProcessingException, because of this exception return not 500 but 401 response.
Didn't remove check in findOrderItems, because of it doesn't allow to place an empty order

Elena-Bruyako commented on src/main/java/org/example/bookshop/controller/OrderController.java:
```diff
@@ -0,0 +1,81 @@
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Update order status",
+            description = "Update order status")
+    public void update(@RequestBody @Valid OrderStatusRequestDto requestDto,
```
```suggestion
    public OrderDto update(@RequestBody @Valid OrderStatusRequestDto requestDto,
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/controller/OrderController.java:
```diff
@@ -0,0 +1,81 @@
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Get OrderItems",
+            description = "Retrieve all OrderItems for a specific order")
+    public Set<OrderItemDto> get(@PathVariable Long orderId,
```
rename method

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,133 @@
+
+    @Override
+    @Transactional
+    public OrderDto add(OrderPurchaseRequestDto requestDto, User user) {
```
rename method

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,133 @@
+
+    @Override
+    public Set<OrderItemDto> get(Long orderId, User user) {
+        findOrderByIdAndUserId(orderId, user.getId());
+        return orderItemRepository.findAllByOrderId(orderId).stream()
```
Don't need to request DB twice, 
see an example
```suggestion
        Order order = orderRepository.findByIdAndUserId(orderId, userId)....
        return orderItemMapper.toOrderItemDtoList(order.getOrderItems());
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,133 @@
+
+    @Override
+    public OrderItemDto getOrderItem(Long orderId, Long itemId, User user) {
+        return orderItemMapper.toDto(
```
```suggestion
         OrderItem item = itemRepository.findByIdAndOrderIdAndUserId(orderItemId, orderId, userId)
                .orElseThrow(()....
         return orderItemMapper.toOrderItemDto(item);
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/Order.java:
```diff
@@ -0,0 +1,61 @@
+    private User user;
+    @Column(nullable = false, unique = true)
+    @Enumerated(EnumType.STRING)
+    private Order.Status status;
```
you can initialize status here

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/Order.java:
```diff
@@ -0,0 +1,61 @@
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
```
you can initialize date now() here

Elena-Bruyako reviewed (changes requested):
see comments

dmarmul committed:
Initialized data and status in model class.
used Query instead EntityGraph and filter by user in repository layer.
Renamed some methods, made method OrderDto update instead of void.

liliia-ponomarenko reviewed (approved):
Great! But please pay attention to my comment ;)

dmarmul committed:
added docker to project

dmarmul committed:
removed docker from project

dmarmul committed:
Delete Dockerfile

dmarmul committed:
Delete docker-compose.yml

dmarmul committed:
Update pom.xml
